plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs

```diff
@@ -180,7 +182,9 @@ compileAlt mustDelay instArgTys (alt, vars, body) = withContextM 3 (sdToTxt $ "C
     -- vars into scope whose body is the body of the case alternative.
     -- See Note [Iterated abstraction and application]
     -- See Note [Case expressions and laziness]
-    GHC.DataAlt _ -> mkIterLamAbsScoped vars (compileExpr body >>= maybeDelay mustDelay)
+    GHC.DataAlt _ -> mkIterLamAbsScoped vars (compiledBody >>= maybeDelay mustDelay)
+    where
+        compiledBody = fromMaybe (compileExpr body) (pure <$> mCompiledBody)
```

Comment:
ugh, this is all a bit subtle, isn't it? There's a secret correspondance going on, where we won't have precompiled versions of the bodies of the cases that have actual lambdas (which is good, because they'd need the variables to be in scope!), so we will always compile them here.

I believe it all works out, it's just... fiddly. I'm not sure how to make it less fiddly.

Reply:
Hmm, here's an idea. How about we have this function return both the delayed _and_ non-delayed bodies. Then we can 

1. run `compileAlt` up front for all the alternatives
2. look at all the non-delayed alternatives to work out if we need to delay
3. if we need to delay, use the delayed versions, otherwise use the non-delayed versions

That wouldn't be too expensive, since we mostly reuse the work: returning the delayed version as well just means making a small change to the non-delayed version.

WDYT?

Reply:
Sounds okay.

Will it affect the purity check? As `compileAlt`'s result is a bit different than `compileExpr body`, the behaviour of `PIR.isPure (const PIR.NonStrict)` might change in some cases?

Reply:
so I'm suggesting that `compileAlt` returns `(compileExpr body, delay $ compileExpr body)` (well not quite, since the delay goes inside the lambda for alternatives with arguments, but you get the idea). Then `isPure` on the first component *will* do the right thing.

Importantly, the reason we check whether there are any variables is just a cheap way of short-cutting what would happen if we compiled the body: if the alternative has variables, then it will turn into a lambda, which will be considered pure.